### DISCUSSION TITLE
Silence gem installation output when executing livecheck

### DISCUSSION THREAD

**Discussion by cesarcoatl** [Homebrew/discussions#1729] [Everyday usage] [answered]
I've created a `~/.brew_livecheck_watchlist` file which includes [`rectangle`](https://github.com/Homebrew/homebrew-cask/blob/9b1e4b11e0e909d33306e2caf309d78b3d701833/Casks/rectangle.rb) among the list of casks and formulae I'm tracking.

The behavior I've noticed is that I get the following output whenever I run `brew livecheck` for the first time after `brew` is updated:

```bash
$ brew livecheck
pyenv : 2.0.1 ==> 2.0.1
Fetching gem metadata from https://rubygems.org/.........
Using concurrent-ruby 1.1.9
Using public_suffix 4.0.6
Using ast 2.4.2
Using bindata 2.4.10
Using msgpack 1.4.2
Using bundler 1.17.3
Using byebug 11.1.3
Using connection_pool 2.2.5
Using diff-lcs 1.4.4
Using docile 1.4.0
Using zeitwerk 2.4.2
Using hpricot 0.8.6
Using minitest 5.14.4
Using mime-types-data 3.2021.0225
Using net-http-digest_auth 1.4.1
Using mini_portile2 2.5.3
Using racc 1.5.2
Using rubyntlm 0.6.3
Using webrick 1.7.0
Using webrobots 0.1.2
Using mustache 1.1.1
Using parallel 1.20.1
Using plist 3.6.0
Using rack 2.2.3
Using rainbow 3.0.0
Using rdiscount 2.2.0.2
Using regexp_parser 2.1.1
Using rexml 3.2.5
Using rspec-support 3.10.2
Using ruby-progressbar 1.11.0
Using unicode-display_width 2.0.0
Using ruby-macho 2.5.1
Using simplecov-html 0.12.3
Using simplecov_json_formatter 0.1.3
Using sorbet-runtime-stub 0.2.0
Using warning 1.2.0
Using i18n 1.8.10
Using tzinfo 2.0.4
Using addressable 2.7.0
Using parser 3.0.1.1
Using elftools 1.1.3
Using bootsnap 1.7.5
Using net-http-persistent 4.0.1
Using unf_ext 0.0.7.7
Using activesupport 6.1.3.2
Using parallel_tests 3.7.0
Using patchelf 1.3.0
Using rspec-core 3.10.1
Using rspec-expectations 3.10.1
Using rspec-mocks 3.10.2
Using rubocop-ast 1.7.0
Using simplecov 0.21.2
Using unf 0.1.4
Using ronn 0.7.3
Using rspec-github 2.3.1
Using rspec-retry 0.6.2
Using nokogiri 1.11.7 (x86_64-darwin)
Using rspec-its 1.3.0
Using rspec 3.10.0
Using rubocop 1.17.0
Fetching rspec_junit_formatter 0.4.1
Using simplecov-cobertura 1.4.2
Using mime-types 3.3.1
Using rspec-wait 0.0.9
Using rubocop-performance 1.11.3
Using rubocop-rails 2.11.0
Using rubocop-rspec 2.4.0
Using rubocop-sorbet 0.6.2
Using domain_name 0.5.20190701
Using http-cookie 1.0.4
Using mechanize 2.8.1
Installing rspec_junit_formatter 0.4.1
Bundle complete! 32 Gemfile dependencies, 71 gems now installed.
Bundled gems are installed into `../../usr/local/Homebrew/Library/Homebrew/vendor/bundle`
rectangle : 0.47,52 ==> 0.47,52
tunnelbear : 4.1.6,1614795843 ==> 4.1.6,1614795843
```

My gut feeling is that this is due to the `sparkle` strategy, but I was wondering if that could be done `--quiet`ly.

Thanks!

**Comment by cho-m** [comment, marked as answer]
> My gut feeling is that this is due to the `sparkle` strategy

You are correct. The output is due to `sparkle` strategy, which runs `Homebrew.install_bundler_gems!` due to needing `nokogiri`: https://github.com/Homebrew/brew/blob/master/Library/Homebrew/livecheck/strategy/sparkle.rb#L60-L61

I believe this was added to avoid the situation were Bundler gems were not yet installed or out-of-date, which leads in `nokogiri` load issue: https://github.com/Homebrew/discussions/discussions/1299

The output may be from `bundle install` command, though there may be some other parts of code that print output too: https://github.com/Homebrew/brew/blob/master/Library/Homebrew/utils/gems.rb#L156

---

> but I was wondering if that could be done `--quiet`ly.

I am not aware of a way to disable this by some argument/flag in current Homebrew.

You might be able to avoid this by running an update prior to livecheck, 
e.g. **EDIT**: maybe also `brew update` so that it isn't triggered somewhere else.
```
brew install-bundler-gems &>/dev/null
brew livecheck
```

---

The proper way to silence it would be using `bundle install --quiet`, but it requires a code change in Homebrew itself.

e.g. For sledgehammer approach of always silencing those lines:
```diff
diff --git a/Library/Homebrew/utils/gems.rb b/Library/Homebrew/utils/gems.rb
index 0eaada3c8..a9b11e2da 100644
--- a/Library/Homebrew/utils/gems.rb
+++ b/Library/Homebrew/utils/gems.rb
@@ -153,7 +153,7 @@ module Homebrew

       # for some reason sometimes the exit code lies so check the output too.
       bundle_installed = if bundle_check_failed || bundle_check_output.include?("Install missing gems")
-        if system bundle, "install"
+        if system bundle, "install", "--quiet"
           true
         else
           message = <<~EOS
```

However, if the output is still wanted for certain commands, then it will require some extra work to add a way of toggling this behavior on/off.

**EDIT:** Another option may be finding a way to redirect stdout/stderr when calling `Homebrew.install_bundler_gems!` from Sparkle strategy.

**Reply by nandahkrishna** [reply]
We had a PR (Homebrew/brew#11479) to redirect bundler output to `stderr`, but it had to be reverted (Homebrew/brew#11500). If the suggestion in the revert PR is implemented, we could redirect the bundler output from `stderr` to `/dev/null` during a `livecheck` run to silence the gem installation.

**Reply by cesarcoatl** [reply]
>we could redirect the bundler output from stderr to /dev/null during a livecheck run to silence the gem installation

This definitely sounds like the behavior I considered; just silence it for certain commands like `livecheck`.

Thanks,

**Reply by cho-m** [reply]
With Homebrew/brew#11599 merged, `nokogiri` dependency has been removed along with the requirement to run bundler commands.

This should probably fix the issue with `livecheck` from original comment.

---

Should also allow users on M1 macs to run `brew livecheck` without resorting to Rosetta.

**Reply by cesarcoatl** [reply]
Thanks, @cho-m!

**Reply by cho-m** [reply]
@thecesrom 
As a note, if you still see an issue, it may require Homebrew/brew#11610, which deals with a couple other places `nokogiri` is used other that directly inside Sparkle livecheck.

**Reply by cesarcoatl** [reply]
Noted. Thanks again, @cho-m .
